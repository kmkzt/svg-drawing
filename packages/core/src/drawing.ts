import { Renderer, RendererOption } from './renderer'
import { Path, Point, Command, COMMAND_TYPE } from './svg'
import { BezierCurve } from './bezier'
import { throttle } from './throttle'
import { getPassiveOptions } from './shared/getPassiveOptions'

export interface DrawingOption extends RendererOption {
  penColor?: string
  penWidth?: number
  close?: boolean
  curve?: boolean
  delay?: number
  fill?: string
}

export class SvgDrawing extends Renderer {
  public penColor: string
  public penWidth: number
  public fill: string
  public curve: boolean
  public close: boolean
  public delay: number
  public bezier: BezierCurve
  private _drawPath: Path | null
  private _drawMoveThrottle: this['drawMove']
  private _listenerOption: ReturnType<typeof getPassiveOptions>
  private _clearPointListener: (() => void) | null
  private _clearMouseListener: (() => void) | null
  private _clearTouchListener: (() => void) | null
  constructor(
    el: HTMLElement,
    {
      penColor,
      penWidth,
      curve,
      close,
      delay,
      fill,
      ...rendOpt
    }: DrawingOption = {}
  ) {
    super(el, { ...rendOpt })
    /**
     * Setup parameter
     */
    this.penColor = penColor ?? '#000'
    this.penWidth = penWidth ?? 1
    this.curve = curve ?? true
    this.close = close ?? false
    this.delay = delay ?? 20
    this.fill = fill ?? 'none'
    this.bezier = new BezierCurve()
    this._drawPath = null
    this._listenerOption = getPassiveOptions(false)
    this._clearPointListener = null
    this._clearMouseListener = null
    this._clearTouchListener = null

    /**
     * Setup listener
     */
    this._handleStart = this._handleStart.bind(this)
    this._handleDraw = this._handleDraw.bind(this)
    this._handleDrawForTouch = this._handleDrawForTouch.bind(this)
    this._handleEnd = this._handleEnd.bind(this)
    this._drawMoveThrottle = throttle(this.drawMove, this.delay)
    this.on()
  }

  public clear(): void {
    this.paths = []
    this.update()
  }

  public undo(): void {
    this.paths.pop()
    this.update()
  }

  public changeDelay(delay: number): void {
    this.delay = delay
    this._drawMoveThrottle = throttle(this.drawMove, this.delay)
  }

  public on(): void {
    this.off()

    if (window.PointerEvent) {
      this._setupPointEventListener()
    } else {
      this._setupMouseEventListener()
    }
    if ('ontouchstart' in window) {
      this._setupTouchEventListener()
    }
  }

  public off(): void {
    ;[
      this._clearPointListener,
      this._clearMouseListener,
      this._clearTouchListener,
    ].map((clear) => {
      if (!clear) return
      clear()
      clear = null
    })
  }

  public drawStart(): void {
    if (this._drawPath) return
    this._drawPath = this._createDrawPath()
    this.addPath(this._drawPath)
  }

  public drawMove(x: number, y: number): void {
    if (!this._drawPath) return
    const po: [number, number] = [x - this.left, y - this.top]
    this._addDrawPoint(po)
    if (
      (this._drawPath.attrs.strokeWidth &&
        +this._drawPath.attrs.strokeWidth !== this.penWidth) ||
      this._drawPath.attrs.stroke !== this.penColor
    ) {
      this._drawPath = this._createDrawPath()
      this._addDrawPoint(po)
      this.addPath(this._drawPath)
    }
    this.update()
  }

  public drawEnd(): void {
    if (this.close && this._drawPath) {
      this._drawPath.commands.push(new Command(COMMAND_TYPE.CLOSE))
    }
    this._drawPath = null
    this.update()
  }

  private _addDrawPoint(po: [number, number]) {
    if (!this._drawPath) return
    const commands = this._drawPath.commands
    if (
      commands.length === 0 ||
      commands[commands.length - 1].type === COMMAND_TYPE.CLOSE
    ) {
      commands.push(new Command(COMMAND_TYPE.MOVE, po))
      return
    }

    if (!this.curve || commands.length < 2) {
      commands.push(new Command(COMMAND_TYPE.LINE, po))
      return
    }

    const p1 =
      commands.length === 2
        ? commands[commands.length - 2].point
        : commands[commands.length - 3].point
    const p2 = commands[commands.length - 2].point
    const p3 = commands[commands.length - 1].point

    if (!p1 || !p2 || !p3) {
      commands.push(new Command(COMMAND_TYPE.LINE, po))
      return
    }

    const p4 = new Point(po[0], po[1])
    commands[commands.length - 1] = this.bezier.createCommand(p1, p2, p3, p4)
    commands.push(this.bezier.createCommand(p2, p3, p4, p4))
  }

  private _createDrawPath(): Path {
    this.resizeElement()
    return new Path({
      stroke: this.penColor,
      strokeWidth: String(this.penWidth),
      fill: this.fill,
      strokeLinecap: this.curve ? 'round' : 'mitter',
      strokeLinejoin: this.curve ? 'round' : 'square',
    })
  }

  private _handleStart(ev: TouchEvent | MouseEvent | PointerEvent) {
    ev.preventDefault()
    this.drawStart()
  }

  private _handleEnd(ev: TouchEvent | MouseEvent | PointerEvent) {
    ev.preventDefault()
    this.drawEnd()
  }

  private _handleDrawForTouch(ev: TouchEvent) {
    ev.preventDefault()
    const touch = ev.touches[0]
    this._drawMoveThrottle(touch.clientX, touch.clientY)
  }

  private _handleDraw(ev: MouseEvent | PointerEvent) {
    ev.preventDefault()
    this._drawMoveThrottle(ev.clientX, ev.clientY)
  }

  /**
   * Drawing PointerEvent
   */
  private _setupPointEventListener(): void {
    this.el.addEventListener(
      'pointerdown',
      this._handleStart,
      this._listenerOption
    )
    this.el.addEventListener(
      'pointermove',
      this._handleDraw,
      this._listenerOption
    )
    this.el.addEventListener(
      'pointerleave',
      this._handleEnd,
      this._listenerOption
    )
    this.el.addEventListener(
      'pointercancel',
      this._handleEnd,
      this._listenerOption
    )
    window.addEventListener('pointerup', this._handleEnd, this._listenerOption)

    this._clearPointListener = () => {
      this.el.removeEventListener('pointerdown', this._handleStart)
      this.el.removeEventListener('pointermove', this._handleDraw)
      this.el.removeEventListener('pointerleave', this._handleEnd)
      this.el.addEventListener('pointercancel', this._handleEnd)
      window.removeEventListener('pointerup', this._handleEnd)
    }
  }
  /**
   * Drawing MouseEvent
   */
  private _setupMouseEventListener(): void {
    this.el.addEventListener(
      'mousedown',
      this._handleStart,
      this._listenerOption
    )
    this.el.addEventListener(
      'mousemove',
      this._handleDraw,
      this._listenerOption
    )
    this.el.addEventListener(
      'mouseleave',
      this._handleEnd,
      this._listenerOption
    )
    this.el.addEventListener('mouseout', this._handleEnd, this._listenerOption)
    window.addEventListener('mouseup', this._handleEnd, this._listenerOption)

    this._clearMouseListener = () => {
      this.el.removeEventListener('mousedown', this._handleStart)
      this.el.removeEventListener('mousemove', this._handleDraw)
      this.el.removeEventListener('mouseleave', this._handleEnd)
      this.el.removeEventListener('mouseout', this._handleEnd)
      window.removeEventListener('mouseup', this._handleEnd)
    }
  }

  /**
   * Drawing TouchEvent
   */
  private _setupTouchEventListener(): void {
    this.el.addEventListener(
      'touchstart',
      this._handleStart,
      this._listenerOption
    )
    this.el.addEventListener(
      'touchmove',
      this._handleDrawForTouch,
      this._listenerOption
    )
    this.el.addEventListener('touchend', this._handleEnd, this._listenerOption)
    window.addEventListener(
      'touchcancel',
      this._handleEnd,
      this._listenerOption
    )
    this._clearTouchListener = () => {
      this.el.removeEventListener('touchstart', this._handleStart)
      this.el.removeEventListener('touchmove', this._handleDrawForTouch)
      this.el.removeEventListener('touchend', this._handleEnd)
      window.removeEventListener('touchcancel', this._handleEnd)
    }
  }
}
