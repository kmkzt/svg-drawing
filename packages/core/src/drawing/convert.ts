import { createCommand } from '../svg/command'
import { Point } from '../svg/point'
import { Vector } from '../svg/vector'
import type {
  EventPoint,
  CommandClass,
  PointObject,
  BezierCurveOption,
} from '../types'

export const createLineCommands = (points: EventPoint[]): CommandClass[] =>
  points.map((p, i) =>
    i === 0
      ? createCommand({ type: 'M', values: [p.x, p.y] })
      : createCommand({ type: 'L', values: [p.x, p.y] })
  )

interface GenerateCommandsConverter {
  create: (points: EventPoint[]) => CommandClass[]
}

export class BezierCurve implements GenerateCommandsConverter {
  public ratio: number
  constructor({ ratio }: BezierCurveOption = {}) {
    this.ratio = ratio ?? 0.4
    this.create = this.create.bind(this)
  }

  public genCommand(
    p1: PointObject,
    p2: PointObject,
    p3: PointObject,
    p4: PointObject
  ): CommandClass {
    const value = Vector.fromPoint(
      new Point(p3.x, p3.y).sub(new Point(p2.x, p2.y))
    ).scale(this.ratio).value

    const vPrev = Point.fromVector({
      angle: Vector.fromPoint(new Point(p3.x, p3.y).sub(new Point(p1.x, p1.y)))
        .angle,
      value,
    })

    const vNext = Point.fromVector({
      angle: Vector.fromPoint(new Point(p2.x, p2.y).sub(new Point(p4.x, p4.y)))
        .angle,
      value,
    })

    const cPrev = new Point(p2.x, p2.y).add(vPrev)
    const cNext = new Point(p3.x, p3.y).add(vNext)

    return createCommand({
      type: 'C',
      values: [...cPrev.values, ...cNext.values, p3.x, p3.y],
    })
  }

  public create(points: EventPoint[]): CommandClass[] {
    const commands: CommandClass[] = []
    if (points.length < 3) {
      return createLineCommands(points)
    }
    for (let i = 0; i < points.length; i += 1) {
      if (i === 0) {
        commands.push(
          createCommand({ type: 'M', values: [points[i].x, points[i].y] })
        )
        continue
      }
      commands.push(
        this.genCommand(
          points[i - 2 < 0 ? 0 : i - 2],
          points[i - 1],
          points[i],
          i < points.length - 1 ? points[i + 1] : points[i]
        )
      )
    }
    return commands
  }
}

export const closeCommands = (commands: CommandClass[]): CommandClass[] => [
  ...commands,
  createCommand({ type: 'Z', values: [] }),
]
