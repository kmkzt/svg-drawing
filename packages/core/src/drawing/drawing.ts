import type { DrawingClass } from '..'
import type { EventPoint, DrawFactory, PathClass, SvgClass } from '../types'

/**
 * ### Basic Drawing usage.
 *
 * ```ts
 * import {
 *   Svg,
 *   Drawing,
 *   PenHandler,
 *   Renderer,
 *   BasicDrawFactory,
 * } from '@svg-drawing/core'
 *
 * const el = document.getElementById('draw')
 * const { width, height } = el.getBoundingClientRect()
 *
 * new Drawing(
 *   new Svg({ width, height }),
 *   new BasicDrawFactory({ stroke: '#000' }),
 *   new PenHandler(el),
 *   new Renderer(el).update
 * )
 * ```
 */
export class Drawing implements DrawingClass {
  private _drawPath: PathClass | null
  private _drawPoints: EventPoint[]
  constructor(
    public svg: SvgClass,
    public pathFactory: DrawFactory,
    private update: (svg: SvgClass) => void
  ) {
    /** Setup property */
    this._drawPath = null
    this._drawPoints = []

    /** Setup EventDrawHandler */
    this.start = this.start.bind(this)
    this.dot = this.dot.bind(this)
    this.end = this.end.bind(this)
  }

  public start(): void {
    this._drawPoints = []
    this._drawPath = this.pathFactory.createPath()
    this.svg.addElement(this._drawPath)
  }

  public dot(po: EventPoint): void {
    if (!this._drawPath) return

    this._drawPoints.push(po)
    this._drawPath.setCommands(this.pathFactory.createCommand(this._drawPoints))

    this.update(this.svg)
  }

  public end(): void {
    this._drawPath = null
    this.update(this.svg)
  }
}
