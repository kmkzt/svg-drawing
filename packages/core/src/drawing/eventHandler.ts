import {
  getEventPoint,
  SUPPORT_EVENT_LISTENER_PASSIVE_OPTION,
  SUPPORT_ON_TOUCH_START,
  SUPPORT_POINTER_EVENT,
} from '../event'
import { throttle } from '../throttle'
import type { DrawingClass } from '..'
import type {
  DrawListenerType,
  DrawEventHandler,
  ClearListener,
  EventPoint,
  DrawEventName,
} from '../types'

class OffsetPosition {
  left: number
  top: number
  constructor(private el?: HTMLElement) {
    const { left, top } = el ? el.getBoundingClientRect() : { left: 0, top: 0 }

    this.left = left
    this.top = top
  }

  setElement(el: HTMLElement) {
    this.el = el
  }

  setup(): Array<ClearListener> {
    const el = this.el
    if (!el) return []

    const setOffsetPosition = () => {
      const { left, top } = el.getBoundingClientRect()
      this.left = left
      this.top = top
    }

    setOffsetPosition()
    addEventListener('scroll', setOffsetPosition)
    el.addEventListener('resize', setOffsetPosition)
    return [
      () => {
        removeEventListener('scroll', setOffsetPosition)
        el.removeEventListener('resize', setOffsetPosition)
      },
    ]
  }
}

export abstract class BaseHandler implements DrawEventHandler {
  /** Remove EventList */
  private _clearEventList: Array<ClearListener>
  /** Offset coordinates */
  private _offsetPosition: OffsetPosition

  constructor(
    protected drawing: DrawingClass,
    protected el?: HTMLElement | undefined
  ) {
    // Bind method
    this.on = this.on.bind(this)
    this.off = this.off.bind(this)
    this.setDrawing = this.setDrawing.bind(this)

    // Set offset coordinates
    this._offsetPosition = new OffsetPosition(el)
    this._clearEventList = []
  }

  public get active(): boolean {
    return this._clearEventList.length > 0
  }

  public off() {
    this._clearEventList.map((fn) => fn())
    this._clearEventList = []
  }

  public on() {
    this.off()

    this._clearEventList = [
      ...this._offsetPosition.setup(),
      ...this.setupListener(),
    ]
  }

  public setElement(el: HTMLElement) {
    this.el = el
    this._offsetPosition.setElement(el)

    if (this.active) this.on()
  }

  public setDrawing(drawing: DrawingClass) {
    this.drawing = drawing
    if (this.active) this.on()
  }

  public getPointObjectFromDrawEvent(
    ev: MouseEvent | TouchEvent | PointerEvent
  ): EventPoint {
    const { x, y, pressure } = getEventPoint(ev)

    return {
      x: x - this._offsetPosition.left,
      y: y - this._offsetPosition.top,
      pressure,
    }
  }

  protected abstract setupListener(): Array<ClearListener>
}

export class PencilHandler extends BaseHandler {
  private _drawMoveThrottle: DrawingClass['dot']
  private delay = 20
  /** AddEventListener Options */
  private listenerOption: { passive: boolean } | false
  constructor(drawing: DrawingClass, el?: HTMLElement) {
    super(drawing, el)

    // Bind methods
    this.setDrawing = this.setDrawing.bind(this)
    this._handleStart = this._handleStart.bind(this)
    this._handleMove = this._handleMove.bind(this)
    this._handleEnd = this._handleEnd.bind(this)
    this._drawMoveThrottle = throttle(this.drawing.dot, this.delay).bind(this)

    this.listenerOption = SUPPORT_EVENT_LISTENER_PASSIVE_OPTION
      ? { passive: false }
      : false
  }

  public changeDelay(delay: number): void {
    this.delay = delay
    this._drawMoveThrottle = throttle(this.drawing.dot, this.delay)

    if (this.active) this.on()
  }

  public setDrawing(drawing: DrawingClass) {
    this.drawing = drawing
    this._drawMoveThrottle = throttle(this.drawing.dot, this.delay)

    if (this.active) this.on()
  }

  protected setupListener(): Array<ClearListener> {
    if (SUPPORT_POINTER_EVENT) return this._setupDrawListener('pointer')

    if (SUPPORT_ON_TOUCH_START) return this._setupDrawListener('touch')

    return this._setupDrawListener('mouse')
  }

  private _handleStart(ev: TouchEvent | MouseEvent | PointerEvent) {
    ev.preventDefault()
    this.drawing.start()
  }

  private _handleEnd(ev: TouchEvent | MouseEvent | PointerEvent) {
    ev.preventDefault()
    this.drawing.end()
  }

  private _handleMove(ev: MouseEvent | PointerEvent | TouchEvent) {
    ev.preventDefault()
    this._drawMoveThrottle(this.getPointObjectFromDrawEvent(ev))
  }

  private _setupDrawListener(type: DrawListenerType): Array<() => void> {
    const el = this.el
    if (!el) return []

    const eventMap: Record<
      DrawListenerType,
      {
        start: Array<DrawEventName>
        move: Array<DrawEventName>
        end: Array<DrawEventName>
        flameout: Array<DrawEventName>
      }
    > = {
      pointer: {
        start: ['pointerdown'],
        move: ['pointermove'],
        end: ['pointerleave', 'pointercancel'],
        flameout: ['pointerup'],
      },
      touch: {
        start: ['touchstart'],
        move: ['touchmove'],
        end: ['touchend'],
        flameout: ['touchcancel'],
      },
      mouse: {
        start: ['mousedown'],
        move: ['mousemove'],
        end: ['mouseleave', 'mouseout'],
        flameout: ['mouseup'],
      },
    }

    const { start, move, end, flameout } = eventMap[type]

    const startClear = start.map((evname): ClearListener => {
      el.addEventListener(evname, this._handleStart, this.listenerOption)
      return () => el.removeEventListener(evname, this._handleStart)
    })

    const moveClear = move.map((evname): ClearListener => {
      el.addEventListener(evname, this._handleMove, this.listenerOption)
      return () => el.removeEventListener(evname, this._handleMove)
    })

    const endClear = end.map((evname): ClearListener => {
      el.addEventListener(evname, this._handleEnd, this.listenerOption)
      return () => el.removeEventListener(evname, this._handleEnd)
    })

    const flameoutClear = flameout.map((evname): ClearListener => {
      addEventListener(evname, this._handleEnd, this.listenerOption)
      return () => removeEventListener(evname, this._handleEnd)
    })

    return [...startClear, ...moveClear, ...endClear, ...flameoutClear]
  }
}

export class PenHandler extends BaseHandler {
  private _editing: boolean
  constructor(drawing: DrawingClass, el?: HTMLElement) {
    super(drawing, el)
    this._editing = false
    this._handleProt = this._handleProt.bind(this)
  }

  protected setupListener(): Array<ClearListener> {
    const clearEvent: ClearListener[] = [...this._setupCancelListener()]
    if (SUPPORT_POINTER_EVENT)
      return [...this._setupDrawListener('pointer'), ...clearEvent]

    if (SUPPORT_ON_TOUCH_START)
      return [...this._setupDrawListener('touch'), ...clearEvent]

    return [...this._setupDrawListener('mouse'), ...clearEvent]
  }

  private _handleProt(ev: MouseEvent | PointerEvent | TouchEvent) {
    ev.preventDefault()
    const isFrameIn = this._isContainElement(ev)
    if (!this._editing && isFrameIn) {
      this.drawing.start()
      this.drawing.dot(this.getPointObjectFromDrawEvent(ev))
      this._editing = true
      return
    }

    if (isFrameIn) {
      this.drawing.dot(this.getPointObjectFromDrawEvent(ev))
      return
    }

    // end
    this._editing = false
    this.drawing.end()
  }

  private _isContainElement(
    ev: MouseEvent | PointerEvent | TouchEvent
  ): boolean {
    if (!this.el) return false
    return this.el.contains(ev.target as any)
  }

  private _setupCancelListener() {
    const stopId = setInterval(() => {
      if (!document.hasFocus()) {
        this._editing = false
        this.drawing.end()
      }
    }, 1000)
    return [() => clearInterval(stopId)]
  }

  private _setupDrawListener(type: DrawListenerType): Array<ClearListener> {
    const eventMap: Record<DrawListenerType, DrawEventName> = {
      touch: 'touchstart',
      pointer: 'pointerup',
      mouse: 'mouseup',
    }

    addEventListener(eventMap[type], this._handleProt)
    return [
      () => {
        removeEventListener(eventMap[type], this._handleProt)
      },
    ]
  }
}
