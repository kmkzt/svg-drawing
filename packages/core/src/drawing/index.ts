export * from './drawing'
export * from './convert'
export * from './factory'
