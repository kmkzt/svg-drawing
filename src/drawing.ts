import { Renderer, RendererOption } from './renderer'
import { Path, Point } from './svg'
import { throttle } from './throttle'
import { getPassiveOptions } from './utils/getPassiveOptions'

export interface DrawingOption extends RendererOption {
  penColor?: string
  penWidth?: number
  close?: boolean
  curve?: boolean
  delay?: number
  fill?: string
}

export class SvgDrawing extends Renderer {
  public penColor: string
  public penWidth: number
  public fill: string
  public curve: boolean
  public close: boolean
  public delay: number
  private _drawPath: Path | null
  private _clearPointListener: (() => void) | null
  private _clearMouseListener: (() => void) | null
  private _clearTouchListener: (() => void) | null
  constructor(
    el: HTMLElement,
    {
      penColor,
      penWidth,
      curve,
      close,
      delay,
      fill,
      ...rendOpt
    }: DrawingOption = {}
  ) {
    super(el, { ...rendOpt })
    /**
     * Setup parameter
     */
    this.penColor = penColor ?? '#333'
    this.penWidth = penWidth ?? 1
    this.curve = curve ?? true
    this.close = close ?? false
    this.delay = delay ?? 20
    this.fill = fill ?? 'none'
    this._drawPath = null
    this._clearPointListener = null
    this._clearMouseListener = null
    this._clearTouchListener = null
    this._drawingMove = this._drawingMove.bind(this) // for throttle
    this.on()
  }

  public clear() {
    this.clearPath()
    this.update()
  }

  public undo() {
    this.undoPath()
    this.update()
  }

  public changeDelay(delay: number) {
    this.delay = delay
    this.on()
  }

  public on() {
    this.off()

    if (window.PointerEvent) {
      this._setupPointEventListener()
    } else {
      this._setupMouseEventListener()
    }
    if ('ontouchstart' in window) {
      this._setupTouchEventListener()
    }
  }

  public off() {
    if (this._clearPointListener) {
      this._clearPointListener()
      this._clearPointListener = null
    }
    if (this._clearMouseListener) {
      this._clearMouseListener()
      this._clearMouseListener = null
    }
    if (this._clearTouchListener) {
      this._clearTouchListener()
      this._clearTouchListener = null
    }
  }

  private _drawingStart() {
    this._drawPath = this._createDrawPath()
    this.addPath(this._drawPath)
  }

  private _drawingMove({ x, y }: { x: number; y: number }): void {
    const po = this._createDrawPoint({ x, y })
    this.addCommand(po)
    if (
      !this._drawPath ||
      this._drawPath.strokeWidth !== this.penWidth ||
      this._drawPath.stroke !== this.penColor
    ) {
      this._drawPath = this._createDrawPath().addCommand(po)
      this.addPath(this._drawPath)
    }
    this.update()
  }

  private _createDrawPath(): Path {
    this.resizeElement()
    return new Path({
      close: this.close,
      curve: this.curve,
      stroke: this.penColor,
      strokeWidth: this.penWidth,
      fill: this.fill
    })
  }
  private _createDrawPoint({ x, y }: { x: number; y: number }): Point {
    return new Point(x - this.left, y - this.top)
  }
  /**
   * Drawing MouseEvent
   */
  private _setupPointEventListener() {
    const handleMouse = (cb: (param: { x: number; y: number }) => void) => (
      ev: PointerEvent
    ): void => {
      ev.preventDefault()
      cb({ x: ev.clientX, y: ev.clientY })
    }
    const start = handleMouse(param => {
      this._drawingStart()
      this.el.addEventListener('pointermove', draw, getPassiveOptions(false))
      this.el.addEventListener('pointerup', end, getPassiveOptions(false))
      this.el.addEventListener('pointerleave', end, getPassiveOptions(false))
      this.el.addEventListener('pointercancel', end, getPassiveOptions(false))
    })
    const draw = throttle(handleMouse(this._drawingMove), this.delay)
    const end = handleMouse((param: { x: number; y: number }) => {
      this.el.removeEventListener('pointermove', draw)
      this.el.removeEventListener('pointerup', end)
      this.el.removeEventListener('pointerleave', end)
      this.el.addEventListener('pointercancel', end)
    })
    this.el.addEventListener('pointerdown', start, getPassiveOptions(false))
    this._clearPointListener = () =>
      this.el.removeEventListener('pointerdown', start)
  }

  /**
   * Drawing MouseEvent
   */
  private _setupMouseEventListener() {
    const handleMouse = (cb: (param: { x: number; y: number }) => void) => (
      ev: MouseEvent
    ): void => {
      ev.preventDefault()
      cb({ x: ev.clientX, y: ev.clientY })
    }
    const start = handleMouse(_param => {
      this._drawingStart()
      this.el.addEventListener('mousemove', draw, getPassiveOptions(false))
      this.el.addEventListener('mouseup', end, getPassiveOptions(false))
      this.el.addEventListener('mouseleave', end, getPassiveOptions(false))
    })
    const draw = throttle(handleMouse(this._drawingMove), this.delay)
    const end = handleMouse(_param => {
      this.el.removeEventListener('mousemove', draw)
      this.el.removeEventListener('mouseup', end)
      this.el.removeEventListener('mouseleave', end)
    })
    this.el.addEventListener('mousedown', start, getPassiveOptions(false))
    this._clearMouseListener = () =>
      this.el.removeEventListener('mousedown', start)
  }

  /**
   * Drawing TouchEvent
   */
  private _setupTouchEventListener() {
    const handleTouch = (cb: (param: { x: number; y: number }) => void) => (
      ev: TouchEvent
    ): void => {
      ev.preventDefault()
      const touch = ev.touches[0]
      cb({ x: touch.clientX, y: touch.clientY })
    }
    const start = handleTouch(_param => {
      this._drawingStart()
      this.el.addEventListener('touchmove', draw, getPassiveOptions(false))
      this.el.addEventListener('touchend', end, getPassiveOptions(false))
    })
    const draw = throttle(handleTouch(this._drawingMove), this.delay)
    const end = handleTouch(_param => {
      this.el.removeEventListener('touchmove', draw)
      this.el.removeEventListener('touchend', end)
    })
    this.el.addEventListener('touchstart', start, getPassiveOptions(false))
    this._clearTouchListener = () =>
      this.el.removeEventListener('touchstart', start)
  }
}
