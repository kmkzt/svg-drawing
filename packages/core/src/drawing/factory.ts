import { BezierCurve, closeCommands, createLineCommands } from './convert'
import { Path } from '../svg/path'
import type {
  CreateCommand,
  DrawFactory,
  PathAttributes,
  PathClass,
} from '../types'

export class BasicDrawFactory implements DrawFactory {
  constructor(
    private attrs: PathAttributes,
    private opts: { curve?: boolean; close?: boolean } = {}
  ) {
    this.opts = {
      curve: true,
      close: false,
      ...opts,
    }
  }

  createPath(): PathClass {
    return new Path({
      ...this.attrs,
      ...this.curveAttribute,
    })
  }

  /** @todo Refactor */
  get createCommand(): CreateCommand {
    const createCommand = this.opts.curve
      ? new BezierCurve().create
      : createLineCommands

    return (po) =>
      this.opts.close ? closeCommands(createCommand(po)) : createCommand(po)
  }

  setPathAttributes(attrs: PathAttributes) {
    this.attrs = attrs
  }

  updatePathAttributes(attrs: PathAttributes) {
    this.attrs = {
      ...this.attrs,
      ...attrs,
    }
  }

  changeCurve(curve: boolean) {
    this.opts = {
      ...this.opts,
      curve,
    }
  }

  changeClose(curve: boolean) {
    this.opts = {
      ...this.opts,
      curve,
    }
  }

  get curveAttribute() {
    return {
      strokeLinecap: this.opts.curve ? 'round' : 'mitter',
      strokeLinejoin: this.opts.curve ? 'round' : 'square',
    }
  }
}
