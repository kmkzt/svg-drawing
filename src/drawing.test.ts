import { SvgDrawing } from './drawing'

describe('drawing.ts', () => {
  describe('SvgDrawing', () => {
    it('default', () => {
      const draw: SvgDrawing = new SvgDrawing(document.createElement('div'))
      draw.drawingStart()
      draw.drawingMove({ x: 0, y: 0 })
      draw.drawingMove({ x: 1, y: 1 })
      draw.drawingMove({ x: 2, y: 1 })
      draw.drawingMove({ x: 3, y: 0 })
      draw.drawingEnd()
      expect(draw.toElement()).toMatchSnapshot()
    })
    it('close', () => {
      const draw: SvgDrawing = new SvgDrawing(document.createElement('div'))
      draw.close = true
      draw.drawingStart()
      draw.drawingMove({ x: 0, y: 0 })
      draw.drawingMove({ x: 1, y: 1 })
      draw.drawingMove({ x: 2, y: 1 })
      draw.drawingMove({ x: 3, y: 0 })
      draw.drawingEnd()
      expect(draw.toElement()).toMatchSnapshot()
    })
    it('curve = false', () => {
      const draw: SvgDrawing = new SvgDrawing(document.createElement('div'))
      draw.curve = false
      draw.drawingStart()
      draw.drawingMove({ x: 0, y: 0 })
      draw.drawingMove({ x: 1, y: 1 })
      draw.drawingMove({ x: 2, y: 1 })
      draw.drawingMove({ x: 3, y: 0 })
      draw.drawingEnd()
      expect(draw.toElement()).toMatchSnapshot()
    })
  })
})
