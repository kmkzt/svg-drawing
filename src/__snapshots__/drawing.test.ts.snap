// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawing.ts SvgDrawing close 1`] = `
<svg
  height="0"
  version="1.1"
  width="0"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M 0 0 C 0.2 0.2 0.6 0.8 1 1 C 1.4 1.2 1.6 1.2 2 1 C 2.4 0.8 2.8 0.2 3 0 Z"
    fill="none"
    stroke="#333"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="1"
  />
</svg>
`;

exports[`drawing.ts SvgDrawing curve = false 1`] = `
<svg
  height="0"
  version="1.1"
  width="0"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M 0 0 L 1 1 L 2 1 L 3 0"
    fill="none"
    stroke="#333"
    stroke-linecap="mitter"
    stroke-linejoin="square"
    stroke-width="1"
  />
</svg>
`;

exports[`drawing.ts SvgDrawing default 1`] = `
<svg
  height="0"
  version="1.1"
  width="0"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M 0 0 C 0.2 0.2 0.6 0.8 1 1 C 1.4 1.2 1.6 1.2 2 1 C 2.4 0.8 2.8 0.2 3 0"
    fill="none"
    stroke="#333"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="1"
  />
</svg>
`;
